From bc4eaaa0272a86cbd93255165ab084221f1844a0 Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Tue, 30 Sep 2014 18:11:15 -0500
Subject: [PATCH 1253/1587] iommu/omap: revise check for valid arch_data in
 attach_dev

Commit a894073 (iommu/omap: check for valid archdata in attach_dev)
has added a sanity check to make sure that the archdata is valid in
attach_dev. Revise this check to also check for the name field, which
would be set correctly only fir properly configured client devices.
The check is also moved outside of the locking primitive.

This revision syncs up this check with the code in upstream vanilla
kernel.

Signed-off-by: Suman Anna <s-anna@ti.com>
[zou: Original patch taken from
ti-sdk-am335x-evm-08.00.00.00-Linux-x86-Install.bin]
Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 drivers/iommu/omap-iommu.c |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/iommu/omap-iommu.c b/drivers/iommu/omap-iommu.c
index f118367..e766c1a 100644
--- a/drivers/iommu/omap-iommu.c
+++ b/drivers/iommu/omap-iommu.c
@@ -1260,14 +1260,13 @@ omap_iommu_attach_dev(struct iommu_domain *domain, struct device *dev)
 	int ret = 0;
 	int i;
 
-	spin_lock(&omap_domain->lock);
-
-	if (!arch_data) {
+	if (!arch_data || !arch_data->name) {
 		dev_err(dev, "device doesn't have an associated iommu\n");
-		ret = -EINVAL;
-		goto out;
+		return -EINVAL;
 	}
 
+	spin_lock(&omap_domain->lock);
+
 	/* only a single client device can be attached to a domain */
 	if (omap_domain->attached) {
 		dev_err(dev, "iommu domain is already attached\n");
-- 
1.7.5.4

